text/3127-trim-paths.md

```diff
@@ -0,0 +1,255 @@
+This flag accepts a comma-separated list of values and may be specified multiple times. The valid scopes are:
+
+- `macro` - apply remappings to the expansion of `std::file!()` macro. This is where paths in embedded panic messages come from
+- `debuginfo` - apply remappings to debug information, wherever they may be written to
+- `split-debuginfo-file` - when `split-debuginfo=packed` or `unpacked`, apply remappings to the paths pointing to these split debug information files
+- `diagnostics` - apply remappings to printed compiler diagnostics
+
+## Cargo
+
+`trim-paths` is a profile setting which controls the sanitisation of file paths in compilation outputs. It has three valid options:
+- `0` or `false`: no sanitisation at all
+- `1`: sanitise only the paths in emitted executable or library binaries. It always affects paths from macros such as panic messages, and in debug information
+  only if they will be embedded together with the binary (the default on platforms with ELF binaries, such as Linux and windows-gnu),
+  but will not touch them if they are in separate files (the default on Windows MSVC and macOS). But the path to these separate files are sanitised.
+- `2` or `true`: sanitise paths in all compilation outputs, including compiled executable/library, debug information, and compiler diagnostics.
+
+The default release profile uses option `1`. You can also manually override it by specifying this option in `Cargo.toml`:
+```toml
+[profile.dev]
+trim-paths = 2
+
+[profile.release]
+trim-paths = 0
+```
+
+When a path is in scope for sanitisation, it is handled by the following rules:
+
+1. Path to the source files of the standard and core library (sysroot) will begin with `/rustc/[rustc commit hash]`.
+   E.g. `/home/username/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/result.rs` -> 
+   `/rustc/fe72845f7bb6a77b9e671e6a4f32fe714962cec4/library/core/src/result.rs`
+2. Path to the working directory will be stripped. E.g. `/home/username/crate/src/lib.rs` -> `src/lib.rs`.
+3. Path to packages outside of the working directory will be replaced with `[package name]-[version]`. E.g. `/home/username/deps/foo/src/lib.rs` -> `foo-0.1.0/src/lib.rs`
+
+When a path to the source files of the standard and core library is *not* in scope for sanitisation, the emitted path will depend on if `rust-src` component
+is present. If it is, then the real path pointing to a copy of the source files on your file system will be emitted; if it isn't, then they will
+show up as `/rustc/[rustc commit hash]/library/...` (just like when it is selected for sanitisation). Paths to all other source files will not be affected.
+
+This will not affect any hard-coded paths in the source code.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## `trim-paths` implementation in Cargo
```

Comment:
Reading this entire section, it doesn't seem to me that this functionality needs to live in Cargo. rustc could directly provide a `-C trim-paths` option with this behavior, since AFAICT rustc has all the information needed to do so. Cargo can then just pass through trim-paths to rustc.

Reply:
Is rustc aware of the path to the current package? This is relevant to dependencies living under `$HOME/.cargo/registry`. Currently there is also the possibility (mentioned in unresolved questions) to encode more information about the package to the sanitised path, such as registry name and git commit hash

Reply:
Hmmm, fair point. I *thought* that rustc had the necessary information, but perhaps not.

It might make sense to pass the requisite information into rustc, but then let rustc make the decision for how to use that information. I'd rather not have cargo making that decision.

Reply:
rustc always knows the absolute path to files, but only the build tool can tell which part of it is environment-sensitive. E.g. rustc knows it's using `/home/cbeuw/.cargo/registry/src/github.com-1ecc6299db9ec823/rand_core-0.6.3/src/lib.rs`, but the `/home/cbeuw/.cargo/registry/src/github.com-1ecc6299db9ec823/rand_core-0.6.3` part is up to Cargo, and it would be different if a different registry were used - but rustc doesn't know that.

This "requisite information" is already included in provided mappings in `--remap-path-prefix`. I can't see how a separate argument that lets the build tool pass in "the sensitive part of the current path" would look like. It'd probably be either too build tool-specific, requiring rustc to know about things like Cargo registries or Bazel repositories, or too "free", ultimately being the same as the existing `--remap-path-prefix`